### DISCUSSION TITLE
fungibility of taproot assets minted within one asset group but in multiple batches

### DISCUSSION THREAD

**Discussion by alex-georgiou** [lightninglabs/taproot-assets#727] [General]
Hello,

I'm not sure if this is really an issue with taproot-assets, or if I have misunderstood something.

I am trying to understand how to issue an asset in multiple batches, so that the total circulating supply is increased every time that I call the [MintAsset](https://lightning.engineering/api-docs/api/taproot-assets/mint/mint-asset) API endpoint.

So far I have only managed to perform multiple mints so that I end up with multiple `asset_id`s, all grouped by the same `group_key`. I have not found a way to issue assets over multiple batches, while preserving the same `asset_id` over all mints. Is there any way to do this? **If there is, please let me know, in which case the remainder of my question is irrelevant.**

My understanding is that we can use the `group_key` to get the [circulating supply](https://lightning.engineering/api-docs/api/taproot-assets/taproot-assets/list-groups) and [hot wallet balance](https://lightning.engineering/api-docs/api/taproot-assets/taproot-assets/list-balances) for a group, and we can conceptually think of the group of assets to be "the asset". But if there is no way to perform multiple mints of the same asset _while preserving the same `asset_id`_, then I think this creates some problems with fungibility:

1. When sending an asset, I have to decide which `asset_id`/batch to send from. If I have minted 1000 units of an asset, and then an additional 1000 units of the same asset in the same group but in a different batch, I can't just send 1100 units in one go.

2. When generating an asset address to receive these, this is done by passing an `asset_id` to [NewAddr](https://lightning.engineering/api-docs/api/taproot-assets/taproot-assets/new-addr). Suppose I have minted an asset over two batches. I now wish to transfer some of this asset. I generate an asset address using the first `asset_id`. Can I now transfer any assets minted by the second batch to that address? If not, then the assets minted between the two batches are not really fungible.

Does this problem really exist? Or have I misunderstood something regarding assets, asset groups, and minting? I would very much appreciate it if someone could explain this to me.

P.S. By the way, I have to say I'm really impressed overall with the Taproot Assets Protocol and with all of the Lightning Labs projects. Thank you all for your effort, it really adds a lot value to the Bitcoin ecosystem.

**Comment by guggero** [comment]
Thank you for asking your question as a discussion, I'm sure other people will find it (and hopefully this answer) useful.

> I have not found a way to issue assets over multiple batches, while preserving the same `asset_id` over all mints.

That is correct. A piece of data that is added to the hash that becomes the `asset_id` is the first on-chain input that is spent in the minting transaction. The reason for this is to add global uniqueness to the `asset_id` in order to prevent someone creating an illegitimate clone of an existing asset.
That of course has the consequence (as you correctly observed) that multiple issuances over multiple minting transactions all result in different `asset_id`s. That is fully on purpose and is what the `group_key` exists for: To group those assets into the same group, making them fungible.

> When generating an asset address to receive these, this is done by passing an `asset_id` to [NewAddr](https://lightning.engineering/api-docs/api/taproot-assets/taproot-assets/new-addr).

It is true that the fungibility described above is not yet fully implemented in our RPCs. For example address based sends is currently only possible with a single `asset_id`, making things not fully fungible yet. That missing feature is described here: https://github.com/lightninglabs/taproot-assets/issues/291.
As you can see in the issue linked, it has the `v0.4` milestone, so it is something we want to address/fix with the next release.

**Reply by alex-georgiou** [reply]
Thank you @guggero, your answer is very clear. I will follow #291 and look forward to `v0.4`.
